### PR TITLE
[WIP] Refactor file searching

### PR DESCRIPTION
**Pull request by MattWis**
First of all, it's worth noting that none of the tests work, unless you comment out the `os_breakcheck();` in `vim_findfile`. Is there a reason the event loop would behave oddly during testing?

```
vim_findfile_init
```

This function only gets called in two places (`find_file_in_path_option` and `tags.c`). Those calls shared many of the same arguments, so I got rid of arguments that were common between them, and some logic associated with them. (Created the stopdir string in the function instead of passing it in, set maximum depth to 100 as a compile-time constant, always reuse visited lists.)

I also extracted `vim_create_stopdirs` to create a list of stopdirs from the stopdirs string.

Finally, I pulled the structs defined at the beginning of the file into the header file instead. This made testing easier (possible?), and allowed me to actually type the arguments to `vim_findfile_init` and `vim_findfile`.

```
find_file_in_path_option
```

I removed the `goto theend;` statements in favor of an early return, since there wasn't any cleanup logic at `theend`.

I extracted `search_with_suffixes`, which tries to add parts of the list of suffixes to a file and find it that way. 

But mainly I just unrolled awkwardly laid out control structures. Most notably, removing `did_findfile_init`, turning:

```
int did_findfile_init = FALSE;
for(;;) {
  if (did_findfile_init) {
    //search for file
    did_findfile_init = FALSE
  } else {
    if (condition) {
      //cleanup
      break;
    }
    //call vim_findfile_init
    did_findfile_init = TRUE;
  }
}
```

to

```
while(!condition) {
  //call vim_findfile_init
  //search for file
}
//cleanup
```

Testing

To test, I looked for the spec file that I had written. Is that a decent method? I could alternatively create files in a temporary directory. 

It's also not cross-platform yet. (Is there a `get_path_seperator()` in lua?) 

I use the lua filesystem (lfs) package.

### CODE REVIEW

**Comment by justinmk** [comment]
See also https://github.com/neovim/neovim/pull/1043 and https://github.com/neovim/neovim/pull/1094

I haven't looked at your PR. If you can explain the value, provide thorough tests, and figure out the cross-platform details, it would be a welcome change. But otherwise, the trend tends to be that people send half-implemented refactorings and lose interest because others don't have the time to spell out the details. Unfortunately you may have to be self-motivated for this type of work (though it is appreciated)!

**Comment by MattWis** [comment]
Just to clarify: Is "this change makes this area of the codebase more readable and adds tests" an acceptable way to explain the value? (Put another way, if I put in the work to get this change the rest of the way, will it get merged in? Or dismissed for not adding enough value to merge in?)

**Comment by elmart** [comment]
Please add description of performed refactorings to commit message, so that others can more easily review this.

**Comment by splinterofchaos** [comment]
I looked at the code a bit--there's a lot going on and large portions of code moved, making it hard to follow...  the `:find` utility, in particular, is one huge state machine that has odd interactions with vim's overall state. Look at the regression test, `src/nvim/testdir/test73.in`. It might be a more important validator of these changes than the unit since it is state aware.

> always reuse visited lists

Doesn't that mean that if I `:find file1` and then `:find file2`, having already searched the directory, vim won't find file2?

Travis found a [memory leak](https://travis-ci.org/neovim/neovim/jobs/51728990). The [error on the unit test](https://travis-ci.org/neovim/neovim/jobs/51728989) is odd..., 

```
Error → ./test/unit/helpers.lua @ 113
/home/travis/build/neovim/neovim/test/unit/file_search_spec.lua
./test/unit/helpers.lua:113: declaration specifier expected near 'FileID' at line 2
stack traceback:
	./test/unit/helpers.lua:113: in function 'cimport'
	...ravis/build/neovim/neovim/test/unit/file_search_spec.lua:10: in main chunk
```

It's hard to tell if #1043 was really a great idea since it naively modified the internal state, and when I attempted to write the regression test for upstream, either I got the test wrong or it had a bug. Bram's comment was that `:find` has been the source of many bugs in the past, so changes to it must be made very carefully. Please, if possible, break up the changes into more granular steps. The changes you've listed seem mostly logical enough, but it's hard to know 100%.

**Comment by justinmk** [comment]
> Or dismissed for not adding enough value to merge in? 

As far as I'm concerned any change that is covered thoroughly by tests is almost guaranteed to be merged, if the change is at least somewhat beneficial. Your changes do sound valuable. 

The tensions about refactoring some of these key pieces of the code are:

- Vim has some very specific (and sometimes stupid) quirks that must be preserved. Refactoring without generous test coverage just causes bugs in an area of the code that no one "owns". 
    - @splinterofchaos is perhaps most familiar with this area, so I'm glad he chimed in!
- If bugs are fixed upstream, we won't get those fixes for "free" anymore, so the change had better be worth it :)
- Need to take #1043 into account if @splinterofchaos thinks it is still a possibility.

> did_findfile_init

Why was Vim caching that, and how do your changes make it unnecessary?

**Comment by justinmk** [comment]
> Is there a reason the event loop would behave oddly during testing?

Yes, there is no event loop for the unit tests. You can spin the event loop manually, I think, or you can set globals manually. If it gets too crazy, you could write "functional" tests instead of unit tests (though this PR should have *some* unit tests for granularity).

> To test, I looked for the spec file that I had written. Is that a decent method? I could alternatively create files in a temporary directory.

Up to you.

> It's also not cross-platform yet.

Do you mean the tests only?

File: src/nvim/file_search.c [hunk 1]
```diff
@@ -1453,44 +1330,32 @@ find_file_in_path_option (
...
-        fdip_search_ctx = vim_findfile_init(buf, ff_file_to_find,
-            r_ptr, 100, FALSE, find_what,
-            fdip_search_ctx, FALSE, rel_fname);
-        if (fdip_search_ctx != NULL)
-          did_findfile_init = TRUE;
-        free(buf);
+      if (fdip_search_ctx != NULL) {
+        file_name = vim_findfile(fdip_search_ctx);
+        if (file_name != NULL)
+          return file_name;
```
Comment (splinterofchaos) [inline src/nvim/file_search.c]: Returning here leaves `buf` unfreed.

File: src/nvim/file_search.c [hunk 2]
```diff
@@ -1509,7 +1374,25 @@ find_file_in_path_option (
     }
   }
 
-theend:
-  return file_name;
+  return NULL;
 }
 
+char_u *search_with_suffixes(char_u *suffixes, size_t l, int find_what) {
```
Comment (splinterofchaos) [inline src/nvim/file_search.c]: This function should be marked `static`. Can `suffixes` be `const char_u *`? Please change `l` to `len` for readability.

**Comment by splinterofchaos** [comment]
> Need to take #1043 into account if @splinterofchaos thinks it is still a possibility.

Shouldn't be a problem, actually. The biggest conflict is that I renamed `ff_stack` to `ff_state`. but that won't be hard to resolve. I might be able to dedicate some time this week for reviving an old PR or two, but that won't be the one.

**Comment by MattWis** [comment]
In order of comments:

@elmart: I'm going to break the commit up so that they make more sense.

@splinterofchaos: It will find both files. (At least it did for me.) Each time you call find, a new search context is initialized. (I believe.)

And failing on importing the header is a really weird error. I'll look into it.

@justinmk: Good to know that it provides value (if I deliver).

The control structure for using ```did_findfile_init``` was the same as a very-convoluted while loop. The same commands should now be run in the same order, but the while loop makes it much clearer what the intended result is. There wasn't really any caching, since it was a local variable initialized to ```FALSE```, and never changed until the ```for(;;)``` loop.

And yes, only the tests are not cross-platform. (Should have made that clearer.)

@splinterofchaos: I'll look through to see what else I can mark as static, and I feel pretty stupid that there's a one letter variable name in there. 

If I get this commit in a decent state, I can look into merging in #1043.

File: src/nvim/file_search.c [hunk 3]
```diff
@@ -208,9 +91,10 @@ static char_u e_pathtoolong[] = N_("E854: path too long for completion");
  *
  * Find the file 'filename' in the directory 'path'.
  * The parameter 'path' may contain wildcards. If so only search 'level'
```
Comment (mseeber) [inline src/nvim/file_search.c]: outdated comment, 'level' was removed from siganture

**Comment by mseeber** [comment]
There are some more outdated comments/descriptions on functions you have changed.

**Comment by justinmk** [comment]
Not enough test coverage, and there doesn't seem to be enough interest in this part of the code for a drastic refactoring.
